contents/Documentation/FP concepts.playground/Pages/Higher Kinded Types.xcplaygroundpage/Contents.swift

```diff
@@ -4,11 +4,193 @@
+*/
+func allIntsEqual(_ array: [Int]) -> Bool {
+    guard let first = array.first else { return false }
+    return array.reduce(true) { partial, next in partial && next == first }
+}
+
+func allStringsEqual(_ array: [String]) -> Bool {
+    guard let first = array.first else { return false }
+    return array.reduce(true) { partial, next in partial && next == first }
+}
+/*:
+ Both functions have exactly the same code but operate in different types. In order to remove this duplication, we can rewrite the type signature and the implementation could remain the same:
+ */
+func allEqual<A: Equatable>(_ array: [A]) -> Bool {
+    guard let first = array.first else { return false }
+    return array.reduce(true) { partial, next in partial && next == first }
+}
+/*:
+ Both functions, `allIntsEqual` and `allStringsEqual` can be replaced by `allEqual`, which works for every type `A`, as long as this type conforms to `Equatable`. `A` is known as the **type parameter** and `allEqual` is a function that uses **parametric polymorphism**; i.e. it has *many forms* that depend on the type parameter we use.
+
+ This enables a degree of generic programming where we can create families of functions that work on many types as long as they conform to certain protocols. However, this is not expressive enough. Let us consider the following functions:
+ */
+enum DivideError: Error {
+    case divisionByZero
+}
+
+func divideEither(x: Int, y: Int) -> Either<DivideError, Int> {
+    guard y != 0 else { return .left(.divisionByZero) }
+    return .right(x / y)
+}
+
+func divideValidated(x: Int, y: Int) -> Validated<DivideError, Int> {
+    guard y != 0 else { return .invalid(.divisionByZero) }
+    return .valid(x / y)
+}
+/*:
+ We can see some similarities between these two functions. They are checking that the second argument is not 0 to perform the division and wrap it in a right/valid case. Otherwise, the division cannot be performed and they wrap an error in a left/invalid case. It would be helpful to unify these two functions in a similar way as in the `allEqual` example presented above. Hypothetically, we would like to write something like:
+
+ ```swift
+    func divide<F: ErrorSuccessRepresentable>(x: Int, y: Int) -> F<DivideError, Int> {
+        guard y != 0 else { return .failure(.divisionByZero)
+        return .success(x / y)
+    }
+ ```
+
+ That is, given a type `F` that is able to create success and error values (via the hypothetical `ErrorSuccessRepresentable` protocol), we could write a function that checks if the division can be performed or not, and create the result values accordingly. This would allow us to write generic programs where we can generalize the container types, not only the contained ones.
+
+ Unfortunately, Swift does not support writting the code above since it does not have HKTs support. Can we find a workaround to have this feature?
```

Comment:
```suggestion
 Unfortunately, Swift does not support writing the code above since it does not have HKTs support. Can we find a workaround to have this feature?
```